functions/snippets/src/main/java/PubSub.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// [START functions_tips_infinite_retries]
+// [START functions_tips_retry]
+import com.google.gson.Gson;
+import com.google.gson.JsonObject;
+import java.time.Duration;
+import java.time.LocalDateTime;
+import java.text.ParseException;
+import java.util.Map;
+
+public class PubSub {
+  // Use Gson (https://github.com/google/gson) to parse JSON content.
+  private Gson gsonParser = new Gson();
+
+  // [END functions_tips_infinite_retries]
+  // [START functions_tips_retry]
+  public String retryPubSub(PubSubMessage message) throws Exception, ParseException {
+    JsonObject body = gsonParser.fromJson(message.data, JsonObject.class);
+    
+    boolean retry = body.has("retry") && body.get("retry").getAsBoolean();
+    if (body.has("retry")) {
+      retry = body.get("retry").getAsBoolean();
+    }
+    
+    if (retry) {
+      throw new RuntimeException("Retrying...");
+    } else {
+      throw new RuntimeException("Not retrying...");
```

Comment:
I still don't think this is right. You want to throw an exception in the retry case, which will result in a 500 status to the client (whatever part of the system is delivering the event). That tells it that it may retry. But in the non-retry case, you want to return normally, which will result in a 200 status indicating that the event has been successfully handled. In the [node](https://github.com/GoogleCloudPlatform/nodejs-docs-samples/blob/master/functions/tips/index.js#L166) equivalent, we do `call(err)` in the retry case and plain `call()` in the non-retry case, and I'm assuming that results in equivalent 500/200 behaviour.